drivers/flash/flash_cadence_nand.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2023, Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT cdns_nand
+
+#include "socfpga_system_manager.h"
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/kernel.h>
+
+/* Check if reset property is defined */
+#define CDNS_NAND_RESET_SUPPORT DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
+
+#if CDNS_NAND_RESET_SUPPORT
+#include <zephyr/drivers/reset.h>
+#endif
+
+#include "flash_cadence_nand_ll.h"
+
+#define DEV_CFG(_dev)  ((const struct flash_cadence_nand_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct flash_cadence_nand_data *const)(_dev)->data)
+
+#define FLASH_WRITE_SIZE DT_PROP(DT_INST(0, DT_DRV_COMPAT), block_size)
+
+#ifdef CONFIG_BOARD_INTEL_SOCFPGA_AGILEX5_SOCDK
+#define DFI_CFG_OFFSET 0xFC
+/* To check the DFI register setting for NAND in the System Manager */
+#define DFI_SEL_CHK    (SOCFPGA_SYSMGR_REG_BASE + DFI_CFG_OFFSET)
+#endif
+
+LOG_MODULE_REGISTER(flash_cdns_nand, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_cadence_nand_data {
+	DEVICE_MMIO_NAMED_RAM(nand_reg);
+	DEVICE_MMIO_NAMED_RAM(sdma);
+	/* device info structure */
+	struct cadence_nand_params params;
+	/* Mutex to prevent multiple processes from accessing the same driver api */
+	struct k_mutex nand_mutex;
+#if CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	/* Semaphore to send a signal from an interrupt handler to a thread  */
+	struct k_sem interrupt_sem;
+#endif
+};
+
+struct flash_cadence_nand_config {
+	DEVICE_MMIO_NAMED_ROM(nand_reg);
+	DEVICE_MMIO_NAMED_ROM(sdma);
+#if CDNS_NAND_RESET_SUPPORT
+	/* Reset controller device configuration for Nand*/
+	const struct reset_dt_spec reset;
+	/* Reset controller device configuration for Combo Phy*/
+	const struct reset_dt_spec combo_phy_reset;
+#endif
+#if CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	void (*irq_config)(void);
+#endif
+};
+
+static const struct flash_parameters flash_cdns_parameters = {.write_block_size = FLASH_WRITE_SIZE,
+							      .erase_value = 0xFF};
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+
+struct flash_pages_layout flash_cdns_pages_layout;
+
+void flash_cdns_page_layout(const struct device *nand_dev, const struct flash_pages_layout **layout,
+			    size_t *layout_size)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+
+	flash_cdns_pages_layout.pages_count = nand_param->page_count;
+	flash_cdns_pages_layout.pages_size = nand_param->page_size;
+	*layout = &flash_cdns_pages_layout;
+	*layout_size = 1;
+}
+
+#endif
+
+static int flash_cdns_nand_erase(const struct device *nand_dev, off_t offset, size_t len)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+	int ret;
+
+	if (len == 0) {
+		LOG_ERR("Invalid input parameter for Nand Flash Erase!");
```

Comment:
I think this message is pointless. This is single point in `flash_cdns_nand_erase` where you can get the -EINVAL.
I think the `LOG_ERR` should be reserved for something that is coming from a lower level, that can not be easily identified.
I am quite sure that this can be easily figured out by user without the message.